### PR TITLE
[CINFRA-819] DropIndex operation

### PR DESCRIPTION
**Pull request by apetenchea**
### Scope & Purpose

DropIndex implementation for replication2. Followup for https://github.com/arangodb/arangodb/pull/19699

Remarks
- When dropping an index, it is not necessary to treat unique indexes separately. Dropping an unique index on the leader makes the leader's policy looser than the follower's. Any documents that are inserted during the drop might still be rejected by the leader. Any document that exists in the log, after the drop-index operation, will not be rejected by the follower due to unique-constraint violation.
- Some methods have been refactored in order to reduce code duplication.

- [ ] :hankey: Bugfix
- [x] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [ ] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [x] **integration tests**
  - [ ] **resilience tests**

### CODE REVIEW

File: arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp [hunk 1]
```diff
@@ -473,6 +473,18 @@ auto DocumentFollowerState::handleSnapshotTransfer(
       });
 }
 
+template<class T>
+auto DocumentFollowerState::GuardedData::applyAndRelease(T const& op,
```
Comment (mchacki) [inline arangod/Replication2/StateMachines/Document/DocumentFollowerState.cpp:477]: Like it :+1:

File: arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp [hunk 2]
```diff
@@ -650,14 +650,65 @@ auto DocumentLeaderState::createIndex(
           }
 
           if (auto releaseRes = self->release(logIndex); releaseRes.fail()) {
-            LOG_CTX("a03c7", ERR, self->loggerContext)
+            LOG_CTX("26355", ERR, self->loggerContext)
```
Comment (mchacki) [inline arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp:653]: Why this LogID change?
Not that it is critical as the log cannot be written yet in production

File: arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.cpp [hunk 3]
```diff
@@ -145,19 +145,31 @@ auto DocumentStateShardHandler::getShardMap() -> ShardMap {
...
+    -> Result {
+  auto res = basics::catchToResult([this, shard = std::move(shard),
+                                    index =
+                                        std::move(index)]() mutable -> Result {
+    return _maintenance->executeDropIndex(std::move(shard), std::move(index));
   });
 
-  if (res.fail()) {
-    res = Result{res.errorNumber(),
-                 fmt::format("Error: {} Failed to ensure index on shard {}: {}",
```
Comment (mchacki) [inline arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.cpp:155]: Don't we loose this additional log information now?

I think this is important as the general Index error message is generic, and this adds the context on the shard.

File: arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.cpp [hunk 4]
```diff
@@ -105,12 +106,13 @@ auto MaintenanceActionExecutor::executeModifyCollectionAction(
 auto MaintenanceActionExecutor::executeCreateIndex(
     ShardID shard, std::shared_ptr<VPackBuilder> const& properties,
     std::shared_ptr<methods::Indexes::ProgressTracker> progress) -> Result {
-  DatabaseGuard guard(_vocbase);
```
Comment (mchacki) [inline arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.cpp:108]: You are loosing the Guard here.
This guard makes sure the vocbase pointer is not dropped.

Is there anything else that guarantees this guard?

File: arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.cpp [hunk 5]
```diff
@@ -122,6 +124,25 @@ auto MaintenanceActionExecutor::executeCreateIndex(
   return res;
 }
 
+auto MaintenanceActionExecutor::executeDropIndex(ShardID shard,
+                                                 velocypack::SharedSlice index)
+    -> Result {
+  auto col = _vocbase.lookupCollection(shard);
```
Comment (mchacki) [inline arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.cpp:130]: If my comment on the DataBaseGuard is correct then we need the guard here as well.

File: arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.cpp (hunk 4 above)
Comment (mchacki) [reply arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.cpp:108]: Come to think about it, should the guard actually be a member of the class that is only erased after the action is completed?

File: arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.h [hunk 6]
```diff
@@ -70,12 +72,17 @@ class MaintenanceActionExecutor : public IMaintenanceActionExecutor {
...
+      -> Result override;
   void addDirty() override;
 
  private:
   GlobalLogIdentifier _gid;
   MaintenanceFeature& _maintenanceFeature;
   ServerID _server;
+
+  // The vocbase reference remains valid for the lifetime of the executor.
+  // Replicated logs are stopped before the vocbase is marked as dropped.
```
Comment (mchacki) [inline arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.h:85]: Ah so this means someone outside holds the DatabaseGuard?
if that is true ignore my above comments on the DatabaseGuard.

File: tests/js/server/replication2/replication2-replicated-state-document-store.js [hunk 7]
```diff
@@ -1001,6 +1000,39 @@ const replicatedStateDocumentShardsSuite = function () {
           }
         }
       }
+
+      // Drop the index.
+      assertTrue(collection.dropIndex(index));
```
Comment (mchacki) [inline tests/js/server/replication2/replication2-replicated-state-document-store.js:1005]: Good test :+1:

**Review by mchacki — CHANGES_REQUESTED** [review]
I mostly have questions, if they get good answers happy to move this to approved.

The general Target of this PR is good, just want to double check that all assumptions are okay.

File: arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp (hunk 2 above)
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp:653]: I've had a conflict with another LOG_CTX and I just patched one of them, I forgot which one was first. Are you ok with this change, or should I revert?

File: arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.h (hunk 6 above)
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.h:85]: There is no need for a `DatabaseGuard`. I have added it "out of reflex" before, but we made sure that, when the database is being dropped or the system is shutting down, the vocbase reference only becomes invalid after the replicated logs have been shut down.
This happens in `DatabaseFeature::dropDatabase()`. There, `shutdownReplicatedLogs()` is called before the vocbase is marked as dropped.

Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.h:85]: I am not against using one though, just for the safety of it, but I would rather do it for all the helpers then, in a way which does not interfere with unit tests. I can do this in a later refactoring.

File: arangod/Cluster/DropIndex.cpp [hunk 8]
```diff
@@ -111,3 +119,21 @@ bool DropIndex::first() {
...
+  if (auto state = vocbase->getReplicatedStateById(col.replicatedStateId());
+      state.ok()) {
+    auto leaderState = std::dynamic_pointer_cast<
+        replication2::replicated_state::document::DocumentLeaderState>(
+        state.get()->getLeader());
+    if (leaderState != nullptr) {
+      return leaderState->dropIndex(col, std::move(index)).get();
+    } else {
+      // TODO prevent busy loop and wait for log to become ready (CINFRA-831)
+      std::this_thread::sleep_for(std::chrono::milliseconds{50});
```
[the remark this answers is not in the document]
Comment (apetenchea) [reply arangod/Cluster/DropIndex.cpp:135]: Yes, and your suggestion would work as well. There are other maintenance actions with the same problem, and I would like to fix them in a single PR ([CINFRA-831](https://arangodb.atlassian.net/browse/CINFRA-831). Ideally I want a method like `waitForLogReady(timeout)`.

File: arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.cpp (hunk 3 above)
Comment (apetenchea) [reply arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.cpp:155]: Addressed in https://github.com/arangodb/arangodb/pull/19760/commits/2493f257db51548e5dc338b416c957773ca59478

File: arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp (hunk 2 above)
Comment (mchacki) [reply arangod/Replication2/StateMachines/Document/DocumentLeaderState.cpp:653]: I am fine with the change :+1:

File: arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.cpp (hunk 3 above)
Comment (mchacki) [reply arangod/Replication2/StateMachines/Document/DocumentStateShardHandler.cpp:155]: :+1:

File: arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.h (hunk 6 above)
Comment (mchacki) [reply arangod/Replication2/StateMachines/Document/MaintenanceActionExecutor.h:85]: Okay, I agree we are protected against dropping, so no Guard required.

**Review by mchacki — APPROVED** [review]
LGTM.

All my questions are addressed.

**Pull request merged**